lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/MultipartUploadUtils.java

```diff
@@ -0,0 +1,63 @@
+package org.sagebionetworks.upload.multipart;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.sagebionetworks.repo.model.upload.PartRange;
+
+public class MultipartUploadUtils {
+
+	protected static final int MAX_PARTS_IN_ONE_COMPOSE = 32;
+
+	/**
+	 * Create a part key using the base and part number.
+	 *
+	 * @param baseKey
+	 * @param partNumber
+	 * @return
+	 */
+	public static String createPartKey(String baseKey, long partNumber) {
+		return String.format("%1$s/%2$d", baseKey, partNumber);
+	}
+
+	/**
+	 * Create a part key using the base and part range.
+	 *
+	 * @param baseKey
+	 * @param lowerBound
+	 * @param upperBound
+	 * @return
+	 */
+	public static String createPartKeyFromRange(String baseKey, long lowerBound, long upperBound) {
+		if (lowerBound==upperBound) {
+			return createPartKey(baseKey, lowerBound);
+		}
+		return String.format("%1$s/%2$d-%3$d", baseKey, lowerBound, upperBound);
+	}
+
+	public static List<PartRange> getListOfPartRangesToLookFor(long lowerBound, long upperBound, long maxPartNumber) {
```

Comment:
This method is ugly, but necessary to compose parts as few times as possible. What this method does is

1. Determines the size of the parts that should be merged based on the part that was most recently added (this is usually the size of the part most recently added, but not necessarily e.g. chunks at the end boundary)
2. Determines the maximum size of the new part after merging (product of the current part size and the maximum number of parts that can be composed)
3. Determines which parts are needed based on
 * the position of the part that was just added
 * the size of the new part
 * the total number of parts that the user is uploading
these parts are returned in a list, which the caller can compare to a list of parts that are actually uploaded. If the lists are equal, then a merge can be performed.

Reply:
getPotentialStitchTargets (or something like that)